File: src/input/input.lua
-- input/input.lua
local utils = require 'utils.utils'

local InputControl = utils.make_class()

--- Initializes everything, duh
function InputControl:_init()
  self.mouse_position   = {love.mouse.getPosition()} -- Current mouse position
  self.mouse_movement   = {0, 0} -- Mouse movement since last update
  self.mouse_press_pos  = {{0, 0}, {0, 0}, {0, 0}} -- Where the mouse was when clicked for dragging
  self.mouse_down       = {false, false, false} -- Mouse buttons current down
  self.mouse_press      = {false, false, false} -- If the mouse was clicked this update
  self.mouse_release    = {false, false, false} -- If the mouse was released this update

  -- self.text_enabled    = false -- Enable storing of keyboard text input
  self.text_input       = "" -- Text input since last update
  self.keyboard_down    = {} -- Keys that are pressed right now
  self.keyboard_press   = {} -- Keys that were pressed this update
  self.keyboard_release = {} -- Keys that were released this update
  
  self.mouse_moved  = false -- If the mouse has been moved this update 
  self.key_pressed  = false -- If a key has been pressed  this update
  self.key_released = false -- If a key has been released this update
end

--- Updates the state of the mouse position
function InputControl:update(dt)
  local mx, my = love.mouse.getPosition()
  self.mouse_movement[1] = mx - self.mouse_position[1]
  self.mouse_movement[2] = my - self.mouse_position[2]
  self.mouse_position[1] = mx
  self.mouse_position[2] = my
  if self.mouse_movement[1] ~= 0 or self.mouse_movement[2] ~= 0 then
    self.mouse_moved = true
  else
    self.mouse_moved = false
  end
end

--- Sets all state to 0
function InputControl:clear(dt)
  for k=1,3 do
    self.mouse_press[k] = false
    self.mouse_release[k] = false
  end
  
  for k in pairs(self.keyboard_press) do
    self.keyboard_press[k] = false
  end
  
  for k in pairs(self.keyboard_release) do
    self.keyboard_press[k] = false
  end
  
  self.text_input = ""
  self.key_pressed  = false
  self.key_released = false
end

--- Updates mouse button state
function InputControl:mouse_button(button, x, y, press)
  -- Note: Button should be in the range 1:3
  self.mouse_down[button] = press
  local c = press and self.mouse_press or self.mouse_release
  c[button] = true
  self.mouse_press_pos[button][1] = press and x or 0
  self.mouse_press_pos[button][2] = press and y or 0
end

--- Updates keyboard state
function InputControl:keyboard_button(keycode, press)
  self.keyboard_down[keycode] = press
  local c = press and self.keyboard_press or self.keyboard_release
  c[keycode] = true
  if press then
    self.key_pressed = true
  else
    self.key_released = true
  end
end

--- Buffers text input
function InputControl:add_text_input(text)
  -- if self.text_enabled then
  self.text_input = self.text_input .. text -- OPT
  -- end
end

--- Returns mouse position as a tuple
function InputControl:get_mouse_position()
  return unpack(self.mouse_position)
end

--- Single input global object
local input = InputControl()

return input

File: src/input/input.ts
import { Coordinate, DeltaTime } from '../utils/utils';

type MouseState = [boolean, boolean, boolean];

class InputControl {
    mouse_position: Coordinate; // Position of the mouse
    mouse_movement: Coordinate = [0, 0]; // Movement of the mouse since last update
    mouse_press_pos: [Coordinate, Coordinate, Coordinate] = [[0, 0], [0, 0], [0, 0]]; // Locations of pressing for the 3 main buttons
    mouse_down: MouseState = [false, false, false]; // Mouse button held
    mouse_press: MouseState = [false, false, false]; // Mouse button pressed down this update
    mouse_release: MouseState = [false, false, false]; // Mouse button released this update
    mouse_scroll: Coordinate = [0, 0]; // Mouse scroll [x, y]

    text_input: string = ""; // Text input since last update
    keyboard_down: {[key: string]: boolean} = {}; // Keyboards currently held
    keyboard_press: {[key: string]: boolean} = {}; // Keys pressed this update
    keyboard_release: {[key: string]: boolean} = {}; // Keys released this update

    mouse_moved: boolean = false; // Mouse moved this update
    key_pressed: boolean = false; // Any key pressed this update
    key_released: boolean = false; // Any key released this update

    constructor() {
        this.mouse_position = love.mouse.getPosition();
    }

    update(_: DeltaTime) {
        const [mx, my] = love.mouse.getPosition();
        this.mouse_movement[0] = mx - this.mouse_position[0];
        this.mouse_movement[1] = my - this.mouse_position[1];

        [...this.mouse_movement] = [mx, my];
        this.mouse_moved = (this.mouse_movement[0] || this.mouse_movement[1]) !== 0;
    }

    clear(_: DeltaTime) {
        [...this.mouse_press] = [...this.mouse_release] = [false, false, false];

        for (let key in this.keyboard_press) {
            this.keyboard_press[key] = false;
        }

        for (let key in this.keyboard_release) {
            this.keyboard_release[key] = false;
        }

        this.text_input = "";
        this.key_pressed = this.key_released = false;
        this.mouse_scroll[0] = this.mouse_scroll[1] = 0;
    }

    mouse_button(button: number, x: number, y: number, press: boolean) {
        button--; // Lua is 1-indexed
        this.mouse_down[button] = press;
        const c = press ? this.mouse_press : this.mouse_release;
        c[button] = true;
        [...this.mouse_press_pos[button]] = [press ? x : 0, press ? y : 0];
    }

    mouse_update_scroll(x: number, y: number) {
        this.mouse_scroll[0] += x;
        this.mouse_scroll[1] += y;
    }

    keyboard_button(keycode: string, press: boolean) {
        this.keyboard_down[keycode] = press;
        const c = press ? this.keyboard_press : this.keyboard_release;
        c[keycode] = true;
        press ? this.key_pressed : this.key_released = true;
    }

    add_text_input(text: string) {
        this.text_input += text;
    }
}

export const input = new InputControl();
